### PR TITLE
feat (api): workflow run #538

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/workflow/run_workflow.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"time"
+
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+
+	"github.com/go-gorp/gorp"
+)
+
+func RunFromHook(db gorp.SqlExecutor, w *sdk.Workflow, e *sdk.WorkflowNodeRunHookEvent) (*sdk.WorkflowRun, error) {
```
Comment (sguiheux) [inline engine/api/workflow/run_workflow.go]: doc

File: engine/api/workflow/run_workflow.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	return nil, nil
+}
+
+func ManualRunFromNode(db gorp.SqlExecutor, w *sdk.Workflow, number int64, e *sdk.WorkflowNodeRunManual, nodeID int64) (*sdk.WorkflowRun, error) {
```
Comment (sguiheux) [inline engine/api/workflow/run_workflow.go]: doc

File: engine/api/workflow/run_workflow.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+	lastWorkflowRun, err = LoadRunByID(db, w.ProjectKey, lastWorkflowRun.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return lastWorkflowRun, nil
+}
+
+func ManualRun(db gorp.SqlExecutor, w *sdk.Workflow, e *sdk.WorkflowNodeRunManual) (*sdk.WorkflowRun, error) {
```
Comment (sguiheux) [inline engine/api/workflow/run_workflow.go]: doc

File: engine/api/workflow/schedule_workflow.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		if c.Err() == nil {
+			chanWorkflowNodeRun <- run
+		} else {
+			log.Error("Exiting workflow.dequeueWorkflows: %s", c.Err())
+			return
+		}
+	}
+}
+
+func Scheduler(c context.Context) error {
```
Comment (sguiheux) [inline engine/api/workflow/schedule_workflow.go]: doc

File: engine/api/workflow_run.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+	"github.com/go-gorp/gorp"
+	"github.com/gorilla/mux"
+	"github.com/ovh/cds/engine/api/context"
+	"github.com/ovh/cds/engine/api/workflow"
+	"github.com/ovh/cds/sdk"
+)
+
+func getWorkflowRunsHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) error {
+	//La réponse de votre API sur une collection devra obligatoirement fournir dans les en-têtes HTTP :
```
Comment (sguiheux) [inline engine/api/workflow_run.go]: should be in english :)

File: engine/sql/021_workflows_run.sql [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+-- +migrate Down
+
+DROP TABLE workflow_node_run_job_logs;
+DROP TABLE workflow_node_run_artifact;
+DROP TABLE workflow_node_run_job;
+DROP TABLE workflow_node_run;
+DROP TABLE workflow_run;
+
+qc0qd7vkvq
```
Comment (sguiheux) [inline engine/sql/021_workflows_run.sql]: ???

File: engine/api/cache/cache.go [hunk 7]
```diff
@@ -109,3 +111,10 @@ func Dequeue(queueName string, value interface{}) {
 	}
 	s.Dequeue(queueName, value)
 }
+
```
Comment (yesnault) [inline engine/api/cache/cache.go]: doc

File: engine/api/cache/local.go [hunk 8]
```diff
@@ -125,5 +126,52 @@ func (s *LocalStore) Dequeue(queueName string, value interface{}) {
 		return
 	}
 	json.Unmarshal(b, value)
+	close(elemChan)
+	return
+}
+
```
Comment (yesnault) [inline engine/api/cache/local.go]: doc

File: engine/api/secret/secret.go [hunk 9]
```diff
@@ -189,6 +189,25 @@ func Decrypt(data []byte) ([]byte, error) {
 	return out, nil
 }
 
+func DecryptVariable(v *sdk.Variable) error {
```
Comment (yesnault) [inline engine/api/secret/secret.go]: doc

File: engine/api/workflow/dao_node_run_job.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+func insertWorkflowNodeJobRun(db gorp.SqlExecutor, j *sdk.WorkflowNodeJobRun) error {
+	dbj := JobRun(*j)
+	if err := db.Insert(&dbj); err != nil {
+		return err
+	}
+	j.ID = dbj.ID
+	return nil
+}
+
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_job.go]: doc

File: engine/api/workflow/dao_node_run_job.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+func DeleteNodeJobRuns(db gorp.SqlExecutor, nodeID int64) error {
+	query := `delete from workflow_node_run_job where workflow_node_run_id = $1`
+	_, err := db.Exec(query, nodeID)
+	return err
+}
+
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_job.go]: doc

File: engine/api/workflow/execute_node_run.go [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+	}
+
+	n.Status = newStatus
+	// Save the node run in database
+	if err := updateWorkflowNodeRun(tx, n); err != nil {
+		return sdk.WrapError(fmt.Errorf("Unable to update node id=%d", n.ID), "workflow.execute> Unable to execute node")
+	}
+
+	//Reload the workflow
+	updatedWorkflowRun, err := loadRunByID(tx, workflowRun.ID)
```
Comment (fsamin) [inline engine/api/workflow/execute_node_run.go]: to reprocess the workflow with processWorkflowRun

File: engine/api/cache/redis.go [hunk 13]
```diff
@@ -155,3 +163,47 @@ func (s *RedisStore) Dequeue(queueName string, value interface{}) {
 		log.Warning("redis> Cannot unmarshal %s :%s", queueName, err)
 	}
 }
+
+func (s *RedisStore) DequeueWithContext(queueName string, value interface{}, c context.Context) {
```
Comment (fsamin) [inline engine/api/cache/redis.go]: ok

File: engine/api/workflow/execute_node_run.go [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+		log.Debug("workflow.execute> Unable to take lock on workflow_run ID=%d (%v)", n.WorkflowRunID, err)
+		return nil
+	}
+	if _, err := tx.Exec("select workflow_node_run.* from workflow_node_run where id = $1 for update nowait", n.ID); err != nil {
+		log.Debug("workflow.execute> Unable to take lock on workflow_node_run ID=%d (%v)", n.ID, err)
+		return nil
+	}
+
+	//Reload the workflow run ID to get the node, and the pipeline...
+	workflowRun, errw := loadRunByID(tx, n.WorkflowRunID)
```
Comment (fsamin) [inline engine/api/workflow/execute_node_run.go]: ok

File: engine/api/workflow/process.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+import (
+	"time"
+
+	"fmt"
+
+	"sort"
+
+	"github.com/fsamin/go-dump"
+	"github.com/go-gorp/gorp"
```
Comment (fsamin) [inline engine/api/workflow/process.go]: ok

File: engine/api/workflow/execute_node_job_run_log.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
+package workflow
+
+import (
+	"time"
+
+	"github.com/go-gorp/gorp"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/ovh/cds/sdk"
+)
+
```
Comment (yesnault) [inline engine/api/workflow/execute_node_job_run_log.go]: doc

File: engine/api/workflow/execute_node_job_run_log.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		return nil, err
+	}
+	logs.Done, err = ptypes.TimestampProto(d)
+	if err != nil {
+		return nil, err
+	}
+	return logs, nil
+}
+
```
Comment (yesnault) [inline engine/api/workflow/execute_node_job_run_log.go]: doc

File: engine/api/workflow/execute_node_job_run_log.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+		l.Done, err = ptypes.TimestampProto(d)
+		if err != nil {
+			return nil, err
+		}
+
+		logs = append(logs, *l)
+	}
+	return logs, nil
+}
+
```
Comment (yesnault) [inline engine/api/workflow/execute_node_job_run_log.go]: doc

File: engine/api/workflow/execute_node_job_run_log.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+	if errd != nil {
+		return errd
+	}
+
+	if err := db.QueryRow(query, logs.PipelineBuildJobID, logs.PipelineBuildID, s, m, d, logs.StepOrder, logs.Val).Scan(&logs.Id); err != nil {
+		return err
+	}
+	return nil
+}
+
```
Comment (yesnault) [inline engine/api/workflow/execute_node_job_run_log.go]: doc

File: engine/sql/021_workflows_run.sql (hunk 6 above)
Comment (fsamin) [inline engine/sql/021_workflows_run.sql]: qc0qd7vkvq
test_1

are just data generated in unit tests :)

File: engine/api/workflow/process.go [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+		Start:          time.Now(),
+		Number:         w.Number,
+		SubNumber:      int64(subnumber),
+		WorkflowRunID:  w.ID,
+		WorkflowNodeID: n.ID,
+		Status:         string(sdk.StatusWaiting),
+		Stages:         n.Pipeline.Stages,
+	}
+
+	for i := range run.Stages {
```
Comment (fsamin) [inline engine/api/workflow/process.go]: I just want them all to be waiting

Comment (fsamin) [inline engine/api/workflow/process.go]: finally i let workflow.execute to set status of each stage sequentally.
it means that not triggered stages have empty status

File: engine/api/cache/cache.go (hunk 7 above)
Comment (fsamin) [inline engine/api/cache/cache.go]: done

File: engine/api/cache/local.go (hunk 8 above)
Comment (fsamin) [inline engine/api/cache/local.go]: done

File: engine/api/secret/secret.go (hunk 9 above)
Comment (fsamin) [inline engine/api/secret/secret.go]: done

File: engine/api/workflow/dao_node_run_job.go (hunk 10 above)
Comment (fsamin) [inline engine/api/workflow/dao_node_run_job.go]: done

File: engine/api/workflow/dao_node_run_job.go (hunk 11 above)
Comment (fsamin) [inline engine/api/workflow/dao_node_run_job.go]: done

File: engine/api/workflow/execute_node_job_run_log.go (hunk 17 above)
Comment (fsamin) [inline engine/api/workflow/execute_node_job_run_log.go]: this func is private now

File: engine/api/workflow/execute_node_job_run_log.go (hunk 18 above)
Comment (fsamin) [inline engine/api/workflow/execute_node_job_run_log.go]: this func is private now

File: engine/api/workflow/execute_node_job_run_log.go (hunk 19 above)
Comment (fsamin) [inline engine/api/workflow/execute_node_job_run_log.go]: this func is private now

**Pull request merged**